### PR TITLE
add proxy support

### PR DESCRIPTION
**Pull request by stiliev**

### CODE REVIEW

**Comment by stiliev** [comment]
I tested with Charles proxy on OSX and with Fiddler proxy on Windows.
The way it works is: create a file ".Proxy" with the host name and port of the proxy - for example http://127.0.0.1:8888. This assumes Fiddler or Charles is set to listen on local port 8888.

**Comment by stiliev** [comment]
Not sure if this is going to address https://github.com/Microsoft/vsts-agent/issues/165 
customer is using a bluecoat proxy. may ask him to retest after this is in.

**Comment by bryanmacfarlane** [comment]
correct - wont address #165 - this addresses debugging and unauthenticated proxies - first step

File: src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+
+        public bool IsBypassed(Uri uri)
+        { 
+            return false;
+        }
+
+        private static bool _proxySettingsApplied = false;
+
+        public static void ApplyProxySettings()
+        {
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs]: This looks susceptible to a race condition. Although I don't know whether it's feasible based on the caller logic. The double-check-and-lock pattern would eliminate the concern:
```
if (!_proxySettingsApplied)
{
    lock (_lock)
    {
        if (!Volative.Read(_proxySettingsApplied))
        {
            try
            {
                // logic to set DefaultWebProxy
            }
            finally
            {
                _proxyConfigured = true;
            }
        }
    }
}

File: src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+        public ICredentials Credentials { get; set; }
+
+        public Uri GetProxy(Uri destination) => ProxyAddress;
+
+        public bool IsBypassed(Uri uri)
+        { 
+            return false;
+        }
+
+        private static bool _proxySettingsApplied = false;
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs]: Can you move this field to the top.

File: src/Microsoft.VisualStudio.Services.Agent/Util/IOUtil.cs [hunk 3]
```diff
@@ -78,6 +78,11 @@ public static string GetServiceConfigFilePath()
             return Path.Combine(GetRootPath(), ".Service");
         }
 
+        public static string GetProxyConfigFilePath()
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/Util/IOUtil.cs]: I think after you rebase, all of the files are lowercase.

File: src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+        public static void ApplyProxySettings()
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs]: Curious if an environment variable was considered. I wonder if that would be a more secure route (anticipating authenticated proxies). Then the worker process could remove the auth details.

File: src/Microsoft.VisualStudio.Services.Agent/Util/ApiUtil.cs [hunk 5]
```diff
@@ -63,6 +64,7 @@ public static VssCredentials GetVssCredential(ServiceEndpoint serviceEndpoint)
             ArgUtil.NotNull(serviceEndpoint.Authorization, nameof(serviceEndpoint.Authorization));
             ArgUtil.NotNullOrEmpty(serviceEndpoint.Authorization.Scheme, nameof(serviceEndpoint.Authorization.Scheme));
 
+            WebProxy.ApplyProxySettings();
```
Comment (ericsciple) [inline src/Microsoft.VisualStudio.Services.Agent/Util/ApiUtil.cs]: Why is this needed in both places (CreateConnection and GetVssCredential)? Can you leave a comment in the source code that explains?

File: src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs (hunk 1 above)
Comment (stiliev) [inline src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs]: I think the code is fine to run in parallel - at worst VssHttpMessageHandler.DefaultWebProxy will be set multiple time. Based on the usage, I don't think this is likely to happen.

File: src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs (hunk 2 above)
Comment (stiliev) [inline src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs]: will move

File: src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs (hunk 4 above)
Comment (stiliev) [inline src/Microsoft.VisualStudio.Services.Agent/WebProxy.cs]: as per our talk with @bryanmacfarlane we settled on a config file. we can change it if it becomes a concern when adding auth proxy support.

File: src/Microsoft.VisualStudio.Services.Agent/Util/ApiUtil.cs (hunk 5 above)
Comment (stiliev) [inline src/Microsoft.VisualStudio.Services.Agent/Util/ApiUtil.cs]: will move the call to ApplyProxySettings in Agent.ExecuteCommand and in Worker.RunAsync, which will also make sure there are no race conditions.
